telemetry/stales.go

```diff
@@ -0,0 +1,54 @@
+/*
+ * Flow Playground
+ *
+ * Copyright 2019 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package telemetry
+
+import (
+	"github.com/dapperlabs/flow-playground-api/model"
+	"github.com/dapperlabs/flow-playground-api/server/config"
+	"github.com/getsentry/sentry-go"
+	"github.com/pkg/errors"
+	"time"
+)
+
+// staleDuration is the amount a time before a project is considered stale if not accessed
+var staleDuration = (time.Hour * 24) * time.Duration(config.Playground().StaleProjectDays)
+
+// staleProjectScanner scans for stale projects in the database
+var staleProjectScanner func(stale time.Duration, projs *[]*model.Project) error = nil
+
+func SetStaleProjectScanner(scanner func(stale time.Duration, projs *[]*model.Project) error) {
+	staleProjectScanner = scanner
+}
+
+// StaleProjectCounter returns the number of stale projects
+func StaleProjectCounter() float64 {
```

Comment:
I'm not sure who is supposed to call this function at regular intervals. So the thing is we need to check stale projects probably every day once and then report that metric. Also don't see where we are setting that metric.

Reply:
Every time you access /metrics endpoint prometheus calls the counter function and updates this automatically. If you think this is too often then I could change it to a separate /stales endpoint or something.

Reply:
Yeah that would be to often. I think there are two ways of doing this, one is, we can have a subroutine that is periodically (once a day) triggered to update the metric, and second, to have an endpoint and then configure k8s to periodically call it. In either case, the metric has to be updated using the gauge, you know what I mean? How was that setup before for a "foo" metric that I wrote as an example: https://pkg.go.dev/github.com/prometheus/client_golang/prometheus?utm_source=godoc#example-Gauge